src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala

```diff
@@ -96,19 +98,48 @@ class ErgoWalletActor(settings: ErgoSettings,
     }
   }
 
+  def scanBlock(block: ErgoFullBlock): Unit = {
+    log.info(s"Wallet is going to scan a block ${block.id} at height ${block.height}")
+    val (reg, offReg) =
+      WalletScanLogic.scanBlockTransactions(registry, offChainRegistry, stateContext, walletVars, block)
+    registry = reg
+    offChainRegistry = offReg
+  }
+
   private def scanLogic: Receive = {
     //scan mempool transaction
     case ScanOffChain(tx) =>
       val newWalletBoxes = WalletScanLogic.extractWalletOutputs(tx, None, walletVars)
       val inputs = WalletScanLogic.extractInputBoxes(tx)
       offChainRegistry = offChainRegistry.updateOnTransaction(newWalletBoxes, inputs)
 
+    case ScanInThePast(blockHeight) =>
+      val lastHeight = registry.fetchDigest().height
+      val expectedHeight = lastHeight + 1
+      if (expectedHeight == blockHeight) {
+        historyReader.bestFullBlockAt(blockHeight) match {
+          case Some(block) =>
+            scanBlock(block)
+          case None =>
+          // We may do not have a block if, for example, the blockchain is pruned. This is okay, just skip it.
+        }
+        if (blockHeight < fullHeight) {
+          self ! ScanInThePast(blockHeight + 1)
+        }
+      }
+
     //scan block transactions
     case ScanOnChain(block) =>
-      val (reg, offReg) =
-        WalletScanLogic.scanBlockTransactions(registry, offChainRegistry, stateContext, walletVars, block)
-      registry = reg
-      offChainRegistry = offReg
+      val lastHeight = registry.fetchDigest().height
+      val expectedHeight = lastHeight + 1
+      if (expectedHeight == block.height) {
+        scanBlock(block)
+      } else if (expectedHeight < block.height) {
+        log.warn(s"Wallet: skipped blocks found starting from $expectedHeight, going back to scan them")
+        self ! ScanInThePast(expectedHeight)
+      } else {
+        log.warn(s"Wallet: $expectedHeight > block.height for a block at ${block.height}, blockId: ${block.id}")
+      }
```

Comment:
I suggest to describe all state transitions of the Wallet in details. I cannot get the logic from the code itself. What is `expectedHeight`? Expected by whom?

The wallet is state machine, hence all the states should be described and the logic of all state changes should be also documented.

Reply:
expectedHeight is height of a block expected by the wallet.

Reply:
On wallet as the state machine, do you mean the scan functionality?

Reply:
I mean state machine is characterized by state and how that state changes when something happens (new message arives etc). Without explicit description it often realy hard to understand even for the author after a couple of years has passed. 
It is better to describe it explicitly.

Reply:
Still , what kind of functionality are you talking about ? If scan, then this is quite simple, scan from the past, scan incoming block, or rollback.

Reply:
I mean documenting the whole ErgoWalletActor, as it represent the state machine.
I assume it is pretty clear what the documentation of state machine should be: 1) what are the states 2) what are the transitions

I suggest to write it down explicitly while you still clearly remember it so that others can understand it in future.

Reply:
Out of scope of the PR then. There's a first issue on writing a test for EWA, so let's make an issue about documentation also. Unlikely the design would be about a single state machine though.